src/Microsoft.AspNet.Mvc.TagHelpers/Internal/FileVersionProvider.cs

```diff
@@ -76,35 +77,39 @@ public string AddFileVersionToPath(string path)
                 return path;
             }
 
-            var fileInfo = _fileProvider.GetFileInfo(resolvedPath);
-            if (!fileInfo.Exists)
+            string value;
+            if (!_cache.TryGetValue(path, out value))
             {
-                if (_requestPathBase.HasValue &&
+                var cacheEntryOptions = new MemoryCacheEntryOptions();
+                var fileInfo = _fileProvider.GetFileInfo(resolvedPath);
+                cacheEntryOptions.AddExpirationToken(_fileProvider.Watch(resolvedPath));
+
+                if (!fileInfo.Exists &&
+                    _requestPathBase.HasValue &&
                     resolvedPath.StartsWith(_requestPathBase.Value, StringComparison.OrdinalIgnoreCase))
                 {
                     resolvedPath = resolvedPath.Substring(_requestPathBase.Value.Length);
                     fileInfo = _fileProvider.GetFileInfo(resolvedPath);
+                    cacheEntryOptions.AddExpirationToken(_fileProvider.Watch(resolvedPath));
                 }
 
-                if (!fileInfo.Exists)
+                if (fileInfo.Exists)
+                {
+                    value = QueryHelpers.AddQueryString(path, VersionKey, GetHashForFile(fileInfo));
+                }
+                else
                 {
                     // if the file is not in the current server.
-                    return path;
+                    value = path;
                 }
-            }
 
-            string value;
-            if (!_cache.TryGetValue(path, out value))
-            {
-                value = QueryHelpers.AddQueryString(path, VersionKey, GetHashForFile(fileInfo));
-                var cacheEntryOptions = new MemoryCacheEntryOptions().AddExpirationToken(_fileProvider.Watch(resolvedPath));
-                _cache.Set(path, value, cacheEntryOptions);
+                value = _cache.Set<string>(path, value, cacheEntryOptions);
```

Comment:
Hmm, do we want to care about thread safety? Could end up with a lotttt of extra file watchers, that is unless the `_fileProvider` caches them.